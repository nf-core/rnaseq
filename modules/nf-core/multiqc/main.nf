process MULTIQC {
    label 'process_single'

    conda "${moduleDir}/environment.yml"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://depot.galaxyproject.org/singularity/multiqc:1.20--pyhdfd78af_0' :
        'biocontainers/multiqc:1.20--pyhdfd78af_0' }"

    input:
    path  multiqc_files, stageAs: "?/*"
    path(multiqc_config)
    path(extra_multiqc_config)
    path(multiqc_logo)

    output:
    path "*multiqc_report.html", emit: report
    path "*_data"              , emit: data
    path "*_plots"             , optional:true, emit: plots
    path "versions.yml"        , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    def config = multiqc_config ? "--config $multiqc_config" : ''
    def extra_config = extra_multiqc_config ? "--config $extra_multiqc_config" : ''
    def logo = multiqc_logo ? /--cl-config 'custom_logo: "${multiqc_logo}"'/ : ''
    """
    multiqc \\
        --force \\
        $args \\
        $config \\
        $extra_config \\
        $logo \\
        .

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        multiqc: \$( multiqc --version | sed -e "s/multiqc, version //g" )
    END_VERSIONS
    """

    stub:
    """
    mkdir multiqc_data
    touch multiqc_plots
    touch multiqc_report.html

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        multiqc: \$( multiqc --version | sed -e "s/multiqc, version //g" )
    END_VERSIONS
    """
}
