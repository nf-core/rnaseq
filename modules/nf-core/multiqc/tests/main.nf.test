nextflow_process {

    name "Test Process MULTIQC"
    script "../main.nf"
    process "MULTIQC"

    tag "modules"
    tag "modules_nfcore"
    tag "multiqc"

    test("sarscov2 single-end [fastqc]") {

        when {
            process {
                """
                input[0] = Channel.of(file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/fastqc/test_fastqc.zip', checkIfExists: true))
                input[1] = []
                input[2] = []
                input[3] = []
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert process.out.report[0] ==~ ".*/multiqc_report.html" },
                { assert process.out.data[0] ==~ ".*/multiqc_data" },
                { assert snapshot(process.out.versions).match("multiqc_versions_single") }
            )
        }

    }

    test("sarscov2 single-end [fastqc] [config]") {

        when {
            process {
                """
                input[0] = Channel.of(file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/fastqc/test_fastqc.zip', checkIfExists: true))
                input[1] = Channel.of(file("https://github.com/nf-core/tools/raw/dev/nf_core/pipeline-template/assets/multiqc_config.yml", checkIfExists: true))
                input[2] = []
                input[3] = []
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert process.out.report[0] ==~ ".*/multiqc_report.html" },
                { assert process.out.data[0] ==~ ".*/multiqc_data" },
                { assert snapshot(process.out.versions).match("multiqc_versions_config") }
            )
        }
    }

    test("sarscov2 single-end [fastqc] - stub") {

        options "-stub"

        when {
            process {
                """
                input[0] = Channel.of(file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/fastqc/test_fastqc.zip', checkIfExists: true))
                input[1] = []
                input[2] = []
                input[3] = []
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out.report.collect { file(it).getName() } +
                                process.out.data.collect { file(it).getName() } +
                                process.out.plots.collect { file(it).getName() } +
                                process.out.versions ).match("multiqc_stub") }
            )
        }

    }
}
