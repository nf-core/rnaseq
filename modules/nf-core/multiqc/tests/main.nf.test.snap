{
    "multiqc_versions_single": {
        "content": [
            [
                "versions.yml:md5,d320d4c37e349c5588e07e7a31cd4186"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-02-14T09:28:51.744211298"
    },
    "multiqc_stub": {
        "content": [
            [
                "multiqc_report.html",
                "multiqc_data",
                "multiqc_plots",
                "versions.yml:md5,d320d4c37e349c5588e07e7a31cd4186"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-02-14T09:29:28.847433492"
    },
    "multiqc_versions_config": {
        "content": [
            [
                "versions.yml:md5,d320d4c37e349c5588e07e7a31cd4186"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-02-14T09:29:13.223621555"
    }
}